numba/_str_lower.c

```diff
@@ -0,0 +1,67 @@
+#include <Python.h>
+#include "_pymodule.h"
+#include <stdio.h>
+
+
+/* from https://github.com/python/cpython/blob/235e7b2b0d937ba8b4a9172aa72206187e3e1f54/Objects/unicodectype.c#L13 */
+#define LOWER_MASK 0x08
+#define UPPER_MASK 0x80
+#define EXTENDED_CASE_MASK 0X4000
+
+typedef struct {
+	const int upper;
+	const int lower;
+	const int title;
+	const unsigned char decimal;
+	const unsigned char digit;
+	const unsigned short flags;
+} _PyUnicode_TypeRecord;
+
+#include "unicodetype_db.h"
+
+/*from https://github.com/python/cpython/blob/e8113f51a8bdf33188ee30a1c038a298329e7bfa/Objects/unicodectype.c#L45 */
+static const _PyUnicode_TypeRecord * gettyperecord(char c)
+{
+	int index;
+	if (c >= 0x110000){
+	   index = 0;
+    }
+	else{
+        index = index1[(c>>SHIFT)];
+        index = index2[(index<<SHIFT)+(c&((1<<SHIFT)-1))];
+	}
+    
+    return &_PyUnicode_TypeRecords[index];
+
+}
+
+/* from https://github.com/python/cpython/blob/e8113f51a8bdf33188ee30a1c038a298329e7bfa/Objects/unicodectype.c#L211 */
+int _PyUnicode_ToLower(char ch, char *res)
```

Comment:
I understood your point. But I have a question. The cpython implementation first checks if the input is ascii or unicode and then performs accordingly. But for this PR how should I handle ascii case? We don't have implementation for IsAscii method yet. I can import the functionality for unicode for this PR but then it should be able to handle ascii as well right?

Reply:
The data model for strings in Numba has a `is_ascii` boolean attribute (for this reason) which should allow you to branch to the desired implementation.  (Note that the ASCII branch should be an optimization, since ASCII is a subset of Unicode.)